rust/core-lib/src/whitespace.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate xi_rope;
+
+use xi_rope::Rope;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Indentation {
+    Tabs,
+    Spaces(usize)
+}
+
+impl Indentation {
+    pub fn parse(rope: &Rope) -> Result<Option<Self>, String> {
+        let (lines, mut spaces, mut tabs) = (rope.lines_raw_all(), Vec::new(), false);
+
+        for line in lines {
+            match Indentation::parse_line(&line) {
+                Ok(Some(Indentation::Spaces(size))) => spaces.push(size),
+                Ok(Some(Indentation::Tabs)) => tabs = true,
+                Ok(None) => continue,
+                Err(e) => return Err(e),
+            }
+        }
+
+        match (tabs, spaces.len() > 0) {
+            (true, true) => return Err(String::from("Mixed indentation")),
+            (true, false) => Ok(Some(Indentation::Tabs)),
+            (false, true) => Ok(Some(Indentation::Spaces(extract_count(spaces)))),
+            _ => Ok(None)
+        }
+    }
+
+    /// Detects the indentation on a specific line.
+    /// Parses whitespace until first occurrence of something else
+    /// Tries to ignore lines which just contain whitespace, or comments, 
+    /// however, this is hard to implement effectively without a syntax tree
+    pub fn parse_line(line: &str) -> Result<Option<Self>, String> {
+        let (mut spaces, mut tabs) = (0, 0);
+        for char in line.chars() {
+            match char {
+                ' ' => spaces += 1,
+                '\t' => tabs += 1,
+                '/' | '#' | '\n' | '\r' => {
```

Comment:
I think this might be overthinking it. I'm especially hesitant to try and hardcode comment delimiters.

Here's an alternative approach I would consider:

- for each line, check the first character
- if it's a tab, return `Indentation::Tabs`
- if it's a space, count all the leading spaces and then check if the next character is a tab; if yes err, else return your count
- if it's anything else, continue